pkg/ruler/base/api.go

```diff
@@ -356,77 +356,151 @@ func parseGroupName(params map[string]string) (string, error) {
+	ql := req.Form.Get("labels")
+	if ql != "" {
+		labelStrings := strings.Split(ql, ",")
+		ls = make([]labels.Label, len(labelStrings))
+
+		for i, l := range labelStrings {
+			kv := strings.Split(l, ":")
+			if len(kv) != 0 {
+				ls[i] = labels.Label{
+					Name:  kv[0],
+					Value: kv[1],
+				}
+			}
 		}
 	}
 
-	return userID, namespace, group, nil
+	return Request{
+		UserID:    userID,
+		Namespace: namespace,
+		Group:     group,
+		Labels:    ls,
+	}, nil
+}
+
+func fitlerRuleGroups(rgs rulespb.RuleGroupList, ls []labels.Label) rulespb.RuleGroupList {
+	res := make(rulespb.RuleGroupList, 0)
+	for _, rg := range rgs {
+		nrg := *rg
+		nrg.Rules = make([]*rulespb.RuleDesc, 0)
+		for _, rule := range rg.Rules {
+			for _, la := range rule.Labels {
+				l := labels.Label{
+					Name:  la.Name,
+					Value: la.Value,
+				}
+				if contains(ls, l) {
+					nrg.Rules = append(nrg.Rules, rule)
+					break
+				}
+			}
+		}
+		if len(nrg.Rules) > 0 {
+			res = append(res, &nrg)
+		}
+	}
+
+	return res
+}
+
+func contains(labels []labels.Label, label labels.Label) bool {
+	for _, l := range labels {
+		if l.Name == label.Name && l.Value == label.Value {
+			return true
+		}
+	}
+
+	return false
 }
 
 func (a *API) ListRules(w http.ResponseWriter, req *http.Request) {
 	logger := util_log.WithContext(req.Context(), a.logger)
 
-	userID, namespace, _, err := parseRequest(req, false, false)
+	pr, err := parseRequest(req, false, false)
 	if err != nil {
 		respondError(logger, w, err.Error())
 		return
 	}
 
-	level.Debug(logger).Log("msg", "retrieving rule groups with namespace", "userID", userID, "namespace", namespace)
-	rgs, err := a.store.ListRuleGroupsForUserAndNamespace(req.Context(), userID, namespace)
+	level.Debug(logger).Log("msg", "retrieving rule groups with namespace", "userID", pr.UserID, "namespace", pr.Namespace)
+	rgs, err := a.store.ListRuleGroupsForUserAndNamespace(req.Context(), pr.UserID, pr.Namespace)
 	if err != nil {
 		http.Error(w, err.Error(), http.StatusBadRequest)
 		return
 	}
 
 	if len(rgs) == 0 {
-		level.Info(logger).Log("msg", "no rule groups found", "userID", userID)
+		level.Info(logger).Log("msg", "no rule groups found", "userID", pr.UserID)
 		http.Error(w, ErrNoRuleGroups.Error(), http.StatusNotFound)
 		return
 	}
 
-	err = a.store.LoadRuleGroups(req.Context(), map[string]rulespb.RuleGroupList{userID: rgs})
+	err = a.store.LoadRuleGroups(req.Context(), map[string]rulespb.RuleGroupList{pr.UserID: rgs})
 	if err != nil {
 		http.Error(w, err.Error(), http.StatusBadRequest)
 		return
 	}
 
-	level.Debug(logger).Log("msg", "retrieved rule groups from rule store", "userID", userID, "num_namespaces", len(rgs))
+	level.Debug(logger).Log("msg", "retrieved rule groups from rule store", "userID", pr.UserID, "num_namespaces", len(rgs))
+
+	if len(pr.Labels) > 0 {
+		level.Info(logger).Log("msg", "filtering rule groups with labels", "abels", pr.Labels, "userID", pr.UserID, "namespace", pr.Namespace)
```

Comment:
Not quite sure why we need an additional log line for this; can we not add the filters to the previous log line?

```suggestion
		level.Info(logger).Log("msg", "filtering rule groups with labels", "labels", pr.Labels, "userID", pr.UserID, "namespace", pr.Namespace)
```

Reply:
Not sure I understand what you mean.. Are you suggesting to convert this into a DEBUG log line? The reason why I added the log line was because label filtering happens only when we have labels